From 4415e357fb185f6078181b98f5bc30f412c840d3 Mon Sep 17 00:00:00 2001
From: Dmitry Muhomor <muhomor.dmitry@gmail.com>
Date: Fri, 28 Mar 2025 19:11:30 +0200
Subject: [PATCH] SettingsProvider: add support for further restriction of
 setting access

By default, Settings.{Global,Secure,System} that aren't annotated with `@Readable` are
readable only by preinstalled apps (with some exceptions, see enforceSettingReadable()).

Settings.{Global,Secure} settings are writable only by apps that hold the privileged
WRITE_SECURE_SETTINGS permission. Settings.System are also writable by apps that hold the
WRITE_SETTINGS app-op permission (it's surfaced as "Modify system settings" in the UI).

This commit adds `@Protected` setting annotation, which allows to further restrict settings access
by specifying which system apps are allowed to read and/or write them.
---
 core/java/android/provider/Settings.java      |  80 ++++++++++++--
 .../providers/settings/SettingsProvider.java  | 103 +++++++++++++++++-
 2 files changed, 169 insertions(+), 14 deletions(-)

diff --git a/core/java/android/provider/Settings.java b/core/java/android/provider/Settings.java
index 8f9da9c6f84f..b9a6d5df2ec2 100644
--- a/core/java/android/provider/Settings.java
+++ b/core/java/android/provider/Settings.java
@@ -60,6 +60,7 @@ import android.content.res.Resources;
 import android.database.ContentObserver;
 import android.database.Cursor;
 import android.database.SQLException;
+import android.ext.KnownSystemPackage;
 import android.location.ILocationManager;
 import android.location.LocationManager;
 import android.media.AudioManager;
@@ -3523,7 +3524,9 @@ public final class Settings {
             mAllFields = new ArraySet<>();
             mReadableFieldsWithMaxTargetSdk = new ArrayMap<>();
             getPublicSettingsForClass(callerClass, mAllFields, mReadableFields,
-                    mReadableFieldsWithMaxTargetSdk);
+                    mReadableFieldsWithMaxTargetSdk,
+                    // skip obtaining protectedSettings map, it's not needed for NameValueCache
+                    null);
         }
 
         public boolean putStringForUser(ContentResolver cr, String name, String value,
@@ -4035,9 +4038,54 @@ public final class Settings {
         int maxTargetSdk() default 0;
     }
 
+    @Target({ ElementType.FIELD })
+    @Retention(RetentionPolicy.RUNTIME)
+    private @interface Protected {
+        // read() and readWrite() are required to be empty if immutableValue is non-empty
+        String immutableValue() default "";
+        // Ignored if immutableValue is non-empty
+        boolean restrictReads() default true;
+        // IDs of system packages that are allowed read-only access. Should be empty if
+        // immutableValue is non-empty or restrictReads is false.
+        @KnownSystemPackage.Enum int[] read() default {};
+        // IDs of system packages that are allowed read and write access. Should be empty if
+        // immutableValue is non-empty
+        @KnownSystemPackage.Enum int[] readWrite() default {};
+    }
+
+    /** @hide */
+    public record ProtectedSetting(String key,
+                                   @Nullable String immutableValue,
+                                   boolean restrictReads,
+                                   @KnownSystemPackage.Enum int[] readableBy,
+                                   @KnownSystemPackage.Enum int[] readWritableBy) {
+
+        static ProtectedSetting fromAnnotation(String key, Protected anno) {
+            int[] readableBy = anno.read();
+            int[] readWritableBy = anno.readWrite();
+
+            String immutableValue = anno.immutableValue();
+            if (!immutableValue.isEmpty()) {
+                if (readableBy.length != 0 || readWritableBy.length != 0) {
+                    throw new IllegalArgumentException(key);
+                }
+                return new ProtectedSetting(key, immutableValue, false, readableBy, readWritableBy);
+            }
+
+            boolean restrictReads = anno.restrictReads();
+            if (!restrictReads && readableBy.length != 0) {
+                throw new IllegalArgumentException(key);
+            }
+
+            return new ProtectedSetting(key, null, anno.restrictReads(),
+                    readableBy, readWritableBy);
+        }
+    }
+
     private static <T extends NameValueTable> void getPublicSettingsForClass(
             Class<T> callerClass, Set<String> allKeys, Set<String> readableKeys,
-            ArrayMap<String, Integer> keysWithMaxTargetSdk) {
+            ArrayMap<String, Integer> keysWithMaxTargetSdk,
+            @Nullable ArrayMap<String, ProtectedSetting> protectedSettings) {
         final Field[] allFields = callerClass.getDeclaredFields();
         try {
             for (int i = 0; i < allFields.length; i++) {
@@ -4049,17 +4097,24 @@ public final class Settings {
                 if (!value.getClass().equals(String.class)) {
                     continue;
                 }
-                allKeys.add((String) value);
+                final String key = (String) value;
+                allKeys.add(key);
                 final Readable annotation = field.getAnnotation(Readable.class);
 
                 if (annotation != null) {
-                    final String key = (String) value;
                     final int maxTargetSdk = annotation.maxTargetSdk();
                     readableKeys.add(key);
                     if (maxTargetSdk != 0) {
                         keysWithMaxTargetSdk.put(key, maxTargetSdk);
                     }
                 }
+
+                if (protectedSettings != null) {
+                    final Protected anno = field.getAnnotation(Protected.class);
+                    if (anno != null) {
+                        protectedSettings.put(key, ProtectedSetting.fromAnnotation(key, anno));
+                    }
+                }
             }
         } catch (IllegalAccessException ignored) {
         }
@@ -4282,9 +4337,10 @@ public final class Settings {
 
         /** @hide */
         public static void getPublicSettings(Set<String> allKeys, Set<String> readableKeys,
-                ArrayMap<String, Integer> readableKeysWithMaxTargetSdk) {
+                ArrayMap<String, Integer> readableKeysWithMaxTargetSdk,
+                ArrayMap<String, ProtectedSetting> protectedSettings) {
             getPublicSettingsForClass(System.class, allKeys, readableKeys,
-                    readableKeysWithMaxTargetSdk);
+                    readableKeysWithMaxTargetSdk, protectedSettings);
         }
 
         /**
@@ -7112,9 +7168,10 @@ public final class Settings {
 
         /** @hide */
         public static void getPublicSettings(Set<String> allKeys, Set<String> readableKeys,
-                ArrayMap<String, Integer> readableKeysWithMaxTargetSdk) {
+                ArrayMap<String, Integer> readableKeysWithMaxTargetSdk,
+                ArrayMap<String, ProtectedSetting> protectedSettings) {
             getPublicSettingsForClass(Secure.class, allKeys, readableKeys,
-                    readableKeysWithMaxTargetSdk);
+                    readableKeysWithMaxTargetSdk, protectedSettings);
         }
 
         /**
@@ -18331,14 +18388,15 @@ public final class Settings {
 
         /** @hide */
         public static void getPublicSettings(Set<String> allKeys, Set<String> readableKeys,
-                ArrayMap<String, Integer> readableKeysWithMaxTargetSdk) {
+                ArrayMap<String, Integer> readableKeysWithMaxTargetSdk,
+                ArrayMap<String, ProtectedSetting> protectedSettings) {
             getPublicSettingsForClass(Global.class, allKeys, readableKeys,
-                    readableKeysWithMaxTargetSdk);
+                    readableKeysWithMaxTargetSdk, protectedSettings);
             // Add Global.Wearable keys on watches.
             if (ActivityThread.currentApplication().getApplicationContext().getPackageManager()
                     .hasSystemFeature(PackageManager.FEATURE_WATCH)) {
                 getPublicSettingsForClass(Global.Wearable.class, allKeys, readableKeys,
-                        readableKeysWithMaxTargetSdk);
+                        readableKeysWithMaxTargetSdk, protectedSettings);
             }
         }
 
diff --git a/packages/SettingsProvider/src/com/android/providers/settings/SettingsProvider.java b/packages/SettingsProvider/src/com/android/providers/settings/SettingsProvider.java
index f7871bcd0af5..61259a15a9de 100644
--- a/packages/SettingsProvider/src/com/android/providers/settings/SettingsProvider.java
+++ b/packages/SettingsProvider/src/com/android/providers/settings/SettingsProvider.java
@@ -75,6 +75,7 @@ import android.database.Cursor;
 import android.database.MatrixCursor;
 import android.database.sqlite.SQLiteDatabase;
 import android.database.sqlite.SQLiteQueryBuilder;
+import android.ext.KnownSystemPackages;
 import android.hardware.camera2.utils.ArrayUtils;
 import android.media.AudioManager;
 import android.media.IRingtonePlayer;
@@ -345,27 +346,33 @@ public class SettingsProvider extends ContentProvider {
     private static final Set<String> sReadableSecureSettings = new ArraySet<>();
     private static final ArrayMap<String, Integer> sReadableSecureSettingsWithMaxTargetSdk =
             new ArrayMap<>();
+    private static final ArrayMap<String, Settings.ProtectedSetting> sProtectedSecureSettings =
+            new ArrayMap<>();
     static {
         Settings.Secure.getPublicSettings(sAllSecureSettings, sReadableSecureSettings,
-                sReadableSecureSettingsWithMaxTargetSdk);
+                sReadableSecureSettingsWithMaxTargetSdk, sProtectedSecureSettings);
     }
 
     private static final Set<String> sAllSystemSettings = new ArraySet<>();
     private static final Set<String> sReadableSystemSettings = new ArraySet<>();
     private static final ArrayMap<String, Integer> sReadableSystemSettingsWithMaxTargetSdk =
             new ArrayMap<>();
+    private static final ArrayMap<String, Settings.ProtectedSetting> sProtectedSystemSettings =
+            new ArrayMap<>();
     static {
         Settings.System.getPublicSettings(sAllSystemSettings, sReadableSystemSettings,
-                sReadableSystemSettingsWithMaxTargetSdk);
+                sReadableSystemSettingsWithMaxTargetSdk, sProtectedSystemSettings);
     }
 
     private static final Set<String> sAllGlobalSettings = new ArraySet<>();
     private static final Set<String> sReadableGlobalSettings = new ArraySet<>();
     private static final ArrayMap<String, Integer> sReadableGlobalSettingsWithMaxTargetSdk =
             new ArrayMap<>();
+    private static final ArrayMap<String, Settings.ProtectedSetting> sProtectedGlobalSettings =
+            new ArrayMap<>();
     static {
         Settings.Global.getPublicSettings(sAllGlobalSettings, sReadableGlobalSettings,
-                sReadableGlobalSettingsWithMaxTargetSdk);
+                sReadableGlobalSettingsWithMaxTargetSdk, sProtectedGlobalSettings);
     }
 
     private final Object mLock = new Object();
@@ -1544,6 +1551,8 @@ public class SettingsProvider extends ContentProvider {
         // Make sure the caller can change the settings - treated as secure.
         enforceHasAtLeastOnePermission(Manifest.permission.WRITE_SECURE_SETTINGS);
 
+        checkProtectedSettingAccess(/* isRead */ false, name, sProtectedGlobalSettings);
+
         // Resolve the userId on whose behalf the call is made.
         final int callingUserId = resolveCallingUserIdEnforcingPermissions(requestingUserId);
 
@@ -1829,6 +1838,8 @@ public class SettingsProvider extends ContentProvider {
         // Make sure the caller can change the settings.
         enforceHasAtLeastOnePermission(Manifest.permission.WRITE_SECURE_SETTINGS);
 
+        checkProtectedSettingAccess(/* isRead */ false, name, sProtectedSecureSettings);
+
         // Resolve the userId on whose behalf the call is made.
         final int callingUserId = resolveCallingUserIdEnforcingPermissions(requestingUserId);
 
@@ -1993,6 +2004,8 @@ public class SettingsProvider extends ContentProvider {
             }
         }
 
+        checkProtectedSettingAccess(/* isRead */ false, name, sProtectedSystemSettings);
+
         // Resolve the userId on whose behalf the call is made.
         final int callingUserId = resolveCallingUserIdEnforcingPermissions(runAsUserId);
 
@@ -2256,6 +2269,8 @@ public class SettingsProvider extends ContentProvider {
     }
 
     private void enforceSettingReadable(String settingName, int settingsType, int userId) {
+        checkProtectedSettingAccess(/* isRead */ true, settingName, getProtectedSettings(settingsType));
+
         if (UserHandle.getAppId(Binder.getCallingUid()) < Process.FIRST_APPLICATION_UID) {
             return;
         }
@@ -2298,6 +2313,61 @@ public class SettingsProvider extends ContentProvider {
         }
     }
 
+    private void checkProtectedSettingAccess(boolean isRead, String name, ArrayMap<String, Settings.ProtectedSetting> map) {
+        Settings.ProtectedSetting protSetting = map.get(name);
+        if (protSetting == null) {
+            return;
+        }
+        if (isRead && !protSetting.restrictReads()) {
+            return;
+        }
+        if (Binder.getCallingPid() == Process.myPid()) {
+            // SettingsProvider runs in system_server process
+            return;
+        }
+        String callingPackage = getCallingPackage();
+        if (callingPackage == null) {
+            if (Build.IS_DEBUGGABLE) {
+                if (Binder.getCallingUid() == ROOT_UID) {
+                    Slog.d(LOG_TAG, "allowed root to access protected setting " + protSetting.key());
+                    return;
+                }
+            }
+            throw new SecurityException("callingPackage is null when accessing protected setting " + protSetting.key());
+        }
+        var ksp = KnownSystemPackages.get(requireContext());
+
+        for (int id : protSetting.readWritableBy()) {
+            if (callingPackage.equals(ksp.getById(id))) {
+                return;
+            }
+        }
+        if (isRead) {
+            for (int id : protSetting.readableBy()) {
+                if (callingPackage.equals(ksp.getById(id))) {
+                    return;
+                }
+            }
+        }
+
+        if (ksp.shell.equals(callingPackage)) {
+            // ADB is used for testing
+            Slog.d(LOG_TAG, "allowed shell to access protected setting " + protSetting.key());
+            return;
+        }
+
+        throw new SecurityException(callingPackage + " is not allowed to access protected setting " + protSetting.key());
+    }
+
+    private static ArrayMap<String, Settings.ProtectedSetting> getProtectedSettings(int settingsType) {
+        return switch (settingsType) {
+            case SETTINGS_TYPE_GLOBAL -> sProtectedGlobalSettings;
+            case SETTINGS_TYPE_SECURE -> sProtectedSecureSettings;
+            case SETTINGS_TYPE_SYSTEM -> sProtectedSystemSettings;
+            default -> throw new IllegalArgumentException();
+        };
+    }
+
     /**
      * Check if the target settings key is readable. Reject if the caller app is trying to access a
      * settings key defined in the Settings.Secure, Settings.System or Settings.Global and is not
@@ -3274,6 +3344,33 @@ public class SettingsProvider extends ContentProvider {
             // Upgrade the settings to the latest version.
             UpgradeController upgrader = new UpgradeController(userId);
             upgrader.upgradeIfNeededLocked();
+
+            // Init immutable settings
+            int[] typesToInit = userId == UserHandle.USER_SYSTEM ?
+                    new int[] { SETTINGS_TYPE_GLOBAL, SETTINGS_TYPE_SECURE, SETTINGS_TYPE_SYSTEM, } :
+                    new int[] { SETTINGS_TYPE_SECURE, };
+
+            for (int type : typesToInit) {
+                SettingsState settingsState = SettingsRegistry.this.getSettingsLocked(type, userId);
+                ArrayMap<String, Settings.ProtectedSetting> protectedSettings = getProtectedSettings(type);
+                String typeStr = SettingsState.settingTypeToString(type);
+                if (type == SETTINGS_TYPE_SECURE) {
+                    typeStr += " (user " + userId + ")";
+                }
+                for (int i = 0; i < protectedSettings.size(); ++i) {
+                    Settings.ProtectedSetting setting = protectedSettings.valueAt(i);
+                    String immutableValue = setting.immutableValue();
+                    if (immutableValue == null) {
+                        continue;
+                    }
+                    settingsState.insertSettingLocked(setting.key(), immutableValue,
+                            /* tag */ null, /* makeDefault */ false,
+                            SettingsState.SYSTEM_PACKAGE_NAME);
+
+                    Slog.d(LOG_TAG, typeStr + ": initialized " + setting.key()
+                            + " with immutable value " + immutableValue);
+                }
+            }
             return true;
         }
 
-- 
2.43.0

